pkg/server/ptypes/pipeline_ui.go

```diff
@@ -0,0 +1,408 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package ptypes
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/hashicorp/waypoint/internal/pkg/graph"
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+func UI_PipelineRunTreeFromJobs(jobs []*pb.Job, statusReports []*pb.StatusReport) (*pb.UI_PipelineRunTreeNode, error) {
+	jobIdx := make(map[string]*pb.Job)
+	statusReportIdx := make(map[string]*pb.StatusReport)
+	nodeIdx := make(map[string]*pb.UI_PipelineRunTreeNode)
+	var g graph.Graph
+	var rootId string
+	var rootNode *pb.UI_PipelineRunTreeNode
+
+	// Populate job index
+	for _, job := range jobs {
+		jobIdx[job.Id] = job
+	}
+
+	// Populate status report index
+	for _, statusReport := range statusReports {
+		if d := statusReport.GetDeploymentId(); d != "" {
+			statusReportIdx[d] = statusReport
+		} else if r := statusReport.GetReleaseId(); r != "" {
+			statusReportIdx[r] = statusReport
+		}
+	}
+
+	// Populate node index
+	for _, job := range jobs {
+		var step *pb.Pipeline_Step
+		var statusReport *pb.StatusReport
+
+		if stepOp := job.GetPipelineStep(); stepOp != nil {
+			step = stepOp.Step
+		} else {
+			// This job represents a nested or referenced pipeline
+			// invocation. It doesn’t encode all the necessary
+			// information to infer the step details, so we’ll
+			// create a placeholder step and fill in the details as
+			// we learn more from other jobs in the set.
+			step = &pb.Pipeline_Step{
+				Name:      job.Pipeline.Step,
+				DependsOn: []string{},
+				Kind:      &pb.Pipeline_Step_Pipeline_{},
+			}
+		}
+
+		// Look up the latest status report for this node
+		if d := job.Result.GetDeploy().GetDeployment(); d != nil {
+			statusReport = statusReportIdx[d.Id]
+		}
+		if r := job.Result.GetRelease().GetRelease(); r != nil {
+			statusReport = statusReportIdx[r.Id]
+		}
+
+		// Build the node (children will be populated later)
+		node := &pb.UI_PipelineRunTreeNode{
+			Step:               step,
+			Job:                &pb.Ref_Job{Id: job.Id},
+			State:              uiPipelineRunTreeNodeStateFromJob(job),
+			StartTime:          job.AckTime,
+			CompleteTime:       job.CompleteTime,
+			Application:        job.Application,
+			Workspace:          job.Workspace,
+			Result:             job.Result,
+			LatestStatusReport: statusReport,
+			Children: &pb.UI_PipelineRunTreeNode_Children{
+				Mode:  pb.UI_PipelineRunTreeNode_Children_SERIAL,
+				Nodes: []*pb.UI_PipelineRunTreeNode{},
+			},
+		}
+
+		// Register the node the index
+		nodeIdx[job.Id] = node
+	}
+
+	// Build graph from job `DependsOn` fields
+	for i, job := range jobIdx {
+		g.Add(i)
+
+		for _, j := range job.DependsOn {
+			if _, exists := jobIdx[j]; !exists {
+				// If the referenced job isn’t in the set then
+				// we can safely ignore it. It’s likely a
+				// task-releated ancillary job.
+				continue
+			}
+
+			g.Add(j)
+			g.AddEdge(j, i)
+		}
+	}
+
+	// Remove short cuts from the graph
+	g.TransitiveReduction()
```

Comment:
Can you expand in a doc comment here what this is doing for us?

Reply:
Absolutely, will do.